### PR TITLE
Various model refactorings and improvements

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Various model refactorings and improvements as a preparation to support the all-dependencies file in the scanner. For details see the commit messages.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/445)

### CODE REVIEW

**Comment by sschuberth** [comment]
> Move MergedResultBuildTest from analyzer to model

This should be "MergedResult*s*BuildTest".

> This aligns the name of the test with the name of the file that contains the tested code.

I'm not fully getting this: Isn't the MergedResultsBuilder() class being tested?

---

Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 2 of 3 files at r4, 2 of 3 files at r5, 1 of 2 files at r6, 1 of 2 files at r7, 1 of 2 files at r8, 1 of 2 files at r9, 2 of 4 files at r10.
Review status: 5 of 8 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[model/src/main/kotlin/MergedAnalyzerResult.kt, line 100 at r11](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-L8g9ixw1uaIMSntybfN:-L8g9ixw1uaIMSntybfO:bd6zs71) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/029a49a54d3ed6c3abb42e5a271251c2a7d70835/model/src/main/kotlin/MergedAnalyzerResult.kt#L100)):*
> ```Kotlin
> 
> /**
>  * A representation of a locally cloned source code repository.
> ```

Isn't this more or less also what our WorkingTree class represents?

---

*[model/src/main/kotlin/PackageReference.kt, line 48 at r2](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-L8g4jqg0xq4Nl96_1cR:-L8g4jqg0xq4Nl96_1cS:b-joi5cx) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/c72c0e82a76ae591685320654af334e6e1bda3ea/model/src/main/kotlin/PackageReference.kt#L48)):*
> ```Kotlin
>         val errors: List<String> = emptyList()
> ) : Comparable<PackageReference> {
>     val allDependencies: SortedSet<Identifier>
> ```

Why is this a property instead of a function? According to the rules at https://discuss.kotlinlang.org/t/default-getter-setter-versus-normal-function-in-kotlin/2304/2 I would have rather expected a function.

---

*[model/src/test/kotlin/MergedAnalyzerResultTest.kt, line 68 at r7](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-L8g8cO-3o18Gtxz8CyS:-L8g8cO-3o18Gtxz8CyT:bqw0vf8) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/f20038ad1a9c5e34fd928988183efc1c9d112956/model/src/test/kotlin/MergedAnalyzerResultTest.kt#L68)):*
> ```Kotlin
>                     namespace = "namespace-$i",
>                     name = "package-$i",
>                     version = "$i"
> ```

Why no "version-" prefix here?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-:-L8g7TYR2FUkkxQyXOGt:b-t58v0p)*

**Comment by mnonnenmacher** [comment]
Actually it should be "MergedResultsBuilderTest" :)
Yes, but the class is defined in "MergedAnalyzerResult.kt".

---

Review status: 1 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[model/src/main/kotlin/MergedAnalyzerResult.kt, line 100 at r11](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-L8g9ixw1uaIMSntybfN:-L8g_0CzClbt0KlJnQMn:bp8hccp) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/029a49a54d3ed6c3abb42e5a271251c2a7d70835/model/src/main/kotlin/MergedAnalyzerResult.kt#L100)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Isn't this more or less also what our WorkingTree class represents?
</blockquote></details>

Yes, it's somewhat similar, but smaller and as a data class. Also as an inner class of VersionControlSystem in the downloader module WorkingTree could not be used in the model.

---

*[model/src/main/kotlin/PackageReference.kt, line 48 at r2](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-L8g4jqg0xq4Nl96_1cR:-L8gdDep-1o4HA4-IoqZ:brt4cj0) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/c72c0e82a76ae591685320654af334e6e1bda3ea/model/src/main/kotlin/PackageReference.kt#L48)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Why is this a property instead of a function? According to the rules at https://discuss.kotlinlang.org/t/default-getter-setter-versus-normal-function-in-kotlin/2304/2 I would have rather expected a function.
</blockquote></details>

Done. I have also renamed it to "collect..." to have a verb in the function name, and "collect" was already used elsewhere in a similar case. Also did the same with "analyzerResults" in MergedAnalyzerResult.

---

*[model/src/test/kotlin/MergedAnalyzerResultTest.kt, line 68 at r7](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-L8g8cO-3o18Gtxz8CyS:-L8g_ivb9NyS8KRZTOE_:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/f20038ad1a9c5e34fd928988183efc1c9d112956/model/src/test/kotlin/MergedAnalyzerResultTest.kt#L68)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Why no "version-" prefix here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/heremaps/oss-review-toolkit/445#-:-L8g_4gn0k6Br1o5BY_r:bth2npw)*

File: model/src/main/kotlin/Scope.kt [hunk 1]
```diff
@@ -47,6 +47,10 @@ data class Scope(
          */
         val dependencies: SortedSet<PackageReference>
 ) : Comparable<Scope> {
+    fun collectAllDependencies(): SortedSet<Identifier> = dependencies.map { it.id }.toSortedSet().also { result ->
```
Comment (sschuberth) [inline model/src/main/kotlin/Scope.kt]: Here the return type can be omitted as it's not recursive like in `PackageReference`.

File: model/src/main/kotlin/Project.kt [hunk 2]
```diff
@@ -68,6 +68,10 @@ data class Project(
          */
         val scopes: SortedSet<Scope>
 ) : Comparable<Project> {
+    fun collectAllDependencies(): SortedSet<Identifier> = sortedSetOf<Identifier>().also { result ->
```
Comment (sschuberth) [inline model/src/main/kotlin/Project.kt]: Here the return type can be omitted as it's not recursive like in `PackageReference`.

File: model/src/main/kotlin/MergedAnalyzerResult.kt [hunk 3]
```diff
@@ -64,7 +66,16 @@ data class MergedAnalyzerResult(
...
         val errors: SortedMap<Identifier, List<String>>
-)
+) {
+    /**
+     * Create the individual [AnalyzerResult]s this [MergedAnalyzerResult] was built from.
+     */
+    fun createAnalyzerResults() = projects.map { project ->
+            val allDependencies = project.collectAllDependencies()
+            val projectPackages = packages.filter { allDependencies.contains(it.id) }.toSortedSet()
+            AnalyzerResult(allowDynamicVersions, project, projectPackages, errors[project.id] ?: listOf())
```
Comment (sschuberth) [inline model/src/main/kotlin/MergedAnalyzerResult.kt]: Use `emptyList()` here?

File: model/src/test/kotlin/MergedAnalyzerResultTest.kt [hunk 4]
```diff
@@ -23,75 +23,50 @@ import io.kotlintest.matchers.shouldBe
 import io.kotlintest.specs.WordSpec
 
 class MergedAnalyzerResultTest : WordSpec() {
+
```
Comment (sschuberth) [inline model/src/test/kotlin/MergedAnalyzerResultTest.kt]: Nit: Addition of blank line 😉

File: model/src/test/kotlin/MergedAnalyzerResultTest.kt [hunk 5]
```diff
@@ -49,6 +49,19 @@ class MergedAnalyzerResultTest : WordSpec() {
     val analyzerResult2 = AnalyzerResult(true, project2, sortedSetOf(package1, package2, package3), listOf("error-2"))
 
     init {
+        "MergedAnalyzerResult" should {
```
Comment (sschuberth) [inline model/src/test/kotlin/MergedAnalyzerResultTest.kt]: Sorry if I'm getting confused here, but should this say "MergedResultsBuilder" (like in the test below), as that's the class being tested?

File: model/src/main/kotlin/MergedAnalyzerResult.kt [hunk 6]
```diff
@@ -22,6 +22,7 @@ package com.here.ort.model
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties
 import com.fasterxml.jackson.annotation.JsonProperty
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import java.io.File
```
Comment (sschuberth) [inline model/src/main/kotlin/MergedAnalyzerResult.kt]: Nit: This should be sorted next to the block below.

File: model/src/test/kotlin/MergedAnalyzerResultTest.kt [hunk 7]
```diff
@@ -21,10 +21,12 @@ package com.here.ort.model
 
 import io.kotlintest.matchers.shouldBe
 import io.kotlintest.specs.WordSpec
+import java.io.File
```
Comment (sschuberth) [inline model/src/test/kotlin/MergedAnalyzerResultTest.kt]: Nit: Blank line before this one.

File: model/src/main/kotlin/Scope.kt (hunk 1 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/Scope.kt]: Works, but I would like to keep it for consistency with Project (see below) and PackageReference.

File: model/src/main/kotlin/Project.kt (hunk 2 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/Project.kt]: But then Kotlin infers the type TreeSet<String> which I didn't like.

File: model/src/main/kotlin/MergedAnalyzerResult.kt (hunk 3 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/MergedAnalyzerResult.kt]: Done.

File: model/src/main/kotlin/MergedAnalyzerResult.kt (hunk 6 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/MergedAnalyzerResult.kt]: Done.

File: model/src/test/kotlin/MergedAnalyzerResultTest.kt (hunk 7 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/MergedAnalyzerResultTest.kt]: Done.

**Pull request merged**